### PR TITLE
[SPARK-50478][SQL] Fix StringType matching

### PR DESCRIPTION
**Pull request by jovanm-db**
### What changes were proposed in this pull request?
In `canUpCast` method in `UpCastRule.scala` we match against `StringType`, which does not match collated string, as opposed to matching `_: StringType`. Similarly, in `needsTimeZone` method in `Cast.scala` it is matched against non collated StringType. 

### Why are the changes needed?
Upcasting to collated strings was disabled even when upcasting to non collated strings was not. Likewise, casting from timestamp to collated string did not need time zone. 

### Does this PR introduce _any_ user-facing change?
No. 

### How was this patch tested?
Test added in `DataTypeWriteCompatibilitySuite.scala` and `CastSuiteBase.scala`. 

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: sql/api/src/main/scala/org/apache/spark/sql/types/UpCastRule.scala [hunk 1]
```diff
@@ -40,8 +40,8 @@ private[sql] object UpCastRule {
     case (DateType, TimestampNTZType) => true
     case (TimestampNTZType, TimestampType) => true
     case (TimestampType, TimestampNTZType) => true
-    case (_: AtomicType, StringType) => true
-    case (_: CalendarIntervalType, StringType) => true
+    case (_: AtomicType, _: StringType) => true
```
Comment (stefankandic) [inline sql/api/src/main/scala/org/apache/spark/sql/types/UpCastRule.scala:43]: can we add some tests for upcast [here](https://github.com/apache/spark/blob/400a8d3797bdcc9183576e66e84163e4dc00a662/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuiteBase.scala#L659)?

**Review by stefankandic — APPROVED** [review]
LGTM!

File: sql/api/src/main/scala/org/apache/spark/sql/types/UpCastRule.scala (hunk 1 above)
Comment (mihailom-db) [reply sql/api/src/main/scala/org/apache/spark/sql/types/UpCastRule.scala:43]: Also please make sure that this does not allow casting from int to string directly by calling cast. In general by standard we should do `cast (value as string) collate collation`, but it is a good catch that we need to allow this transition in other operations.

**Comment by mihailom-db** [comment]
Please remove draft and add [SQL] component to the title of the PR.

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @jovanm-db and @stefankandic @mihailom-db for review.

**Pull request closed without merging**